### ISSUE TITLE
Bug: recursor invalid action for NXDOMAIN cache

### ISSUE THREAD

**Issue by magicbear** [PowerDNS/pdns#5246]
- Program: Recursor
 - Issue type: Bug report

### Short description
When using zone-forwards to query a not exists domain, and another the same top tld domain will be directly return NXDOMAIN.

### Environment

 - Operating system: Debian 8.0 jessie
 - Software version: master branch
 - Software source: PowerDNS repository master, compiled yourself

### Steps to reproduce

1. dig an not exists domain
2. dig an domain that normally, will directly return NXDOMAIN

### Actual behaviour
```
dig bt.ai-sky.com @172.16.0.8

; <<>> DiG 9.8.3-P1 <<>> bt.ai-sky.com @172.16.0.8
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 31276
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0

;; QUESTION SECTION:
;bt.ai-sky.com.			IN	A

;; AUTHORITY SECTION:
ai-sky.com.		3082	IN	SOA	lv3ns1.ffdns.net. webmaster.ffdns.net. 2009092868 604800 3600 2419200 3600

;; Query time: 120 msec
;; SERVER: 172.16.0.8#53(172.16.0.8)
;; WHEN: Sun Apr 16 21:56:24 2017
;; MSG SIZE  rcvd: 93


; <<>> DiG 9.8.3-P1 <<>> github.com @172.16.0.8
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 44136
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0

;; QUESTION SECTION:
;github.com.			IN	A

;; AUTHORITY SECTION:
ai-sky.com.		3080	IN	SOA	lv3ns1.ffdns.net. webmaster.ffdns.net. 2009092868 604800 3600 2419200 3600

;; Query time: 12 msec
;; SERVER: 172.16.0.8#53(172.16.0.8)
;; WHEN: Sun Apr 16 21:56:26 2017
;; MSG SIZE  rcvd: 97


Apr 16 21:56:24 [1] bt.ai-sky.com: Wants DNSSEC processing in query for A
Apr 16 21:56:24 [1] bt.ai-sky.com: Looking for CNAME cache hit of 'bt.ai-sky.com|CNAME'
Apr 16 21:56:24 [1] bt.ai-sky.com: No CNAME cache hit of 'bt.ai-sky.com|CNAME' found
Apr 16 21:56:24 [1] bt.ai-sky.com: No cache hit for 'bt.ai-sky.com|A', trying to find an appropriate NS record
Apr 16 21:56:24 [1] bt.ai-sky.com: Cache consultations done, have 1 NS to contact
Apr 16 21:56:24 [1] bt.ai-sky.com.: Nameservers: <empty>(13.05ms)
Apr 16 21:56:24 [1] bt.ai-sky.com: Domain has hardcoded nameserver
Apr 16 21:56:24 [1] bt.ai-sky.com: Resolved '.' NS (empty) to: 114.114.114.114
Apr 16 21:56:24 [1] bt.ai-sky.com: Trying IP 114.114.114.114:53, asking 'bt.ai-sky.com|A'
Apr 16 21:56:24 [1] bt.ai-sky.com: Got 2 answers from (empty) (114.114.114.114), rcode=3 (Non-Existent domain), aa=0, in 21ms
Apr 16 21:56:24 [1] bt.ai-sky.com: accept answer 'ai-sky.com|SOA|lv3ns1.ffdns.net. webmaster.ffdns.net. 2009092868 604800 3600 2419200 3600' from '.' nameservers? 2 YES! - This answer was received from a server we forward to.
Apr 16 21:56:24 [1] bt.ai-sky.com: OPT answer '.' from '.' nameservers
Apr 16 21:56:24 [1] bt.ai-sky.com: determining status after receiving this packet
Apr 16 21:56:24 [1] bt.ai-sky.com: got negative caching indication for name 'bt.ai-sky.com' (accept=1), newtarget='(empty)'
Apr 16 21:56:24 [1] bt.ai-sky.com: status=NXDOMAIN, we are done (have negative SOA)
Apr 16 21:56:24 [1] bt.ai-sky.com: failed (res=3)
Apr 16 21:56:24 1 [1/1] answer to question 'bt.ai-sky.com|A': 0 answers, 0 additional, took 1 packets, 21.057 ms, 0 throttled, 0 timeouts, 0 tcp connections, rcode=3
Apr 16 21:56:26 1 [2/1] question for 'github.com|A' from 172.16.211.101
Apr 16 21:56:26 [2] github.com: Wants DNSSEC processing in query for A
Apr 16 21:56:26 [2] github.com: Looking for CNAME cache hit of 'github.com|CNAME'
Apr 16 21:56:26 [2] github.com: No CNAME cache hit of 'github.com|CNAME' found
Apr 16 21:56:26 [2] github.com: Entire name 'github.com', is negatively cached via '.' & 'com' for another 3080 seconds
Apr 16 21:56:26 1 [2/1] answer to question 'github.com|A': 0 answers, 0 additional, took 0 packets, 0 ms, 0 throttled, 0 timeouts, 0 tcp connections, rcode=3
```

### Other information

config:
```
disable-packetcache=yes
forward-zones-recurse=.=114.114.114.114
packetcache-servfail-ttl=0
quiet=no
trace=on
```

**Comment by Habbie** [comment]
Can you confirm that `root-nx-trust=no` works around this issue? It does for me.

**Comment by pieterlexis** [comment]
I believe this is a duplicate of #5107

**Comment by phonedph1** [comment]
@pieterlexis I think this is something new that is broken in master. a77e6c032249caa264bdba84887e1a0f0b7db1e0 appeared fine, and soon as I pulled latest upstream I was able to reproduce this.

The only reason we got #5107 to work was the CNAME trick to get it to return a ". IN SOA" I believe, nothing else was this simple to break it.

**Comment by phonedph1** [comment]
#5252 sounds like it will fix the particular one.

**Comment by rgacogne** [comment]
Fixed by #5252.
